### PR TITLE
326 include real nyc acs

### PR DESCRIPTION
**Pull request by julialucyhogan**
This PR:
- Adds rails resource, model, & controller for new route "census-tracts" which serve the nyc_acs data
- Replaces Matt's stubbed modal-split model with a new census-tract model
- Moves `transportation/study-area-map/census-tract-data` component to `transportation/study-area-map/census-tract-popup/data`
- Adds `transportation/study-area-map/census-tract-popup/modal-split` component which contains logic to ingest multiple census-tract records required for a single census-tract
- Updates the on-error initializer to re-throw the original intercepted exceptions, which will retain the context/call stack of the originating error ! 

Questions: 
In order to keep the services simple and coupled to the existing data model, getting the full set of data needed for a single census tract requires multiple records are returned. The full "modal split" data is the result of a call to census-tracts with the geoid filter for the census tract applied in the query. Default filter for variables  is defined to include only variables of interest: population, trans_total, trans_auto_total, trans_public_total. I debated if it was better to restructure the data in the backend service to include a single resource per geoid with the variables flattened into, but opted to keep the backend as a CRUD layer to keep the rails implmentation clean -- we would have to define a new type of resource/model, or do some funky stuff within the existing paradigms and so this solution felt easier. Does anyone have a strong preference for a different solution?

Should census-tracts model in backend, frontend, or both be renamed to make it clear that it is a single data variable for the census-tract? `census-tract-variable`?

Should `modal-split` component be renamed? `modal-split-formatter`? `modal-split-renderer`? `modal-split-displayer` ? All seemed clunky, so I didn't make this one a verb, but I do want it to be clear that the job of those component is to correctly format and display data from census-tract records

### CODE REVIEW

**Comment by godfreyyeung** [comment]
We should also use the CensusTract (or CensusTractRecord) Models in `models/transportation-analysis.jtwStudySelection` array. 
I.e. enforce objects in that array to be instances of the CensusTract(Record) Model.

File: frontend/app/components/transportation/study-area-map/census-tract-popup/modal-split.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+   * RecordArray from census-tract-popup/data containing census-tract models for a given geoid
+   */
+  data = [];
+
+  /**
+   *  The 'value' property of the census-tract record for variable = 'population'
+   */
+  @computed('data')
+  get population() {
+    const data = this.get('data');
```
Comment (allthesignals) [inline frontend/app/components/transportation/study-area-map/census-tract-popup/modal-split.js]: Nit: we shouldn't need get anymore, unless in very specific edge-cases - then there'll be a warning message.

File: frontend/app/components/transportation/study-area-map/census-tract-popup/modal-split.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    const populationVariable = data.filter((censusTractVariable) => {return censusTractVariable.variable === 'population'})[0];
```
Comment (allthesignals) [inline frontend/app/components/transportation/study-area-map/census-tract-popup/modal-split.js]: Nit: Destructuring looks nicer: 
```suggestion
    const [populationVariable] = data.filter((censusTractVariable) => {return censusTractVariable.variable === 'population'});
```

TODO: enable better linting rules

File: frontend/app/initializers/on-error.js [hunk 3]
```diff
@@ -7,10 +7,10 @@ const { shouldThrowOnError } = ENV;
 export function initialize() {
   if (shouldThrowOnError) {
     Ember.onerror = function(err) {
-      throw new Error(err);
+      throw err;
     };
     on('error', function(err) {
-      throw new Error(err);
+      throw err;
```
Comment (allthesignals) [inline frontend/app/initializers/on-error.js]: thx

File: frontend/app/models/census-tract.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  @attr('number') moe;
+
+  /**
+   * Returns the human-readable "mode" string for a given census-tract variable
+   */
+  @computed('variable')
+  get mode() {
+    const variable = this.get('variable');
+    return VARIABLE_MODE_LOOKUP[variable] || 'Unknown';
+  }
```
Comment (allthesignals) [inline frontend/app/models/census-tract.js]: Gorgeous.
non-blocking (bikeshedding?): Only tiny possible issue: the human-readable version of an 'Unknown' is represented in multiple places across the app (See https://github.com/NYCPlanning/ceqr-app/pull/344/files#diff-d323bc5d7c3fca3306820fb54a4a72dbR24). Maybe the lookup could be wrapped in a "crosswalk" function that uses `CASE WHEN`, having a default?

File: frontend/app/templates/components/transportation/study-area-map/census-tract-popup.hbs [hunk 5]
```diff
@@ -10,28 +10,17 @@
       </span>
     </div>
     <div class="description">
-      <Transportation::StudyAreaMap::CensusTractData
+      <Transportation::StudyAreaMap::CensusTractPopup::Data
```
Comment (allthesignals) [inline frontend/app/templates/components/transportation/study-area-map/census-tract-popup.hbs]: Love this

File: frontend/tests/integration/components/transportation/study-area-map-test.js [hunk 6]
```diff
@@ -52,10 +52,12 @@ module('Integration | Component | transportation/study-area-map', function(hooks
   });
 
   test('it hovers, displays information', async function(assert) {
-    this.server.create('modal-split', { id: 'test' });
+    const geoid = '1';
+    this.server.create('census-tract', { geoid });
+    this.server.create('census-tract', { geoid });
```
Comment (allthesignals) [inline frontend/tests/integration/components/transportation/study-area-map-test.js]: fwiw you could say this instead: `this.server.createList('census-tract', 2, { geoid });`

File: frontend/tests/unit/models/census-tract-test.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    let trans_auto_total = store.createRecord('census-tract', { variable: 'trans_auto_total' });
+    let trans_public_total = store.createRecord('census-tract', { variable: 'trans_public_total' });
+
+    // Cannot get the Class constant import to work, so testing with magic strings
+    assert.ok(trans_total.mode, "All");
+    //assert.ok(trans_total.mode, VARIABLE_MODE_LOOKUP['trans_public_total']);
+    assert.ok(trans_auto_total.mode, "Automobile");
+    //assert.ok(trans_auto_total.mode, VARIABLE_MODE_LOOKUP['trans_auto_total']);
+    assert.ok(trans_public_total.mode, "Public Transportation");
+    //assert.ok(trans_public_total.mode, VARIABLE_MODE_LOOKUP['trans_public_total']);
```
Comment (allthesignals) [inline frontend/tests/unit/models/census-tract-test.js]: do we need these commented lines of code?

File: frontend/tests/unit/models/census-tract-test.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+module('Unit | Model | census tract', function(hooks) {
+  setupTest(hooks);
+
+  test('it returns correct display mode from variable', function(assert) {
+    let store = this.owner.lookup('service:store');
+    let trans_total = store.createRecord('census-tract', { variable: 'trans_total' });
+    let trans_auto_total = store.createRecord('census-tract', { variable: 'trans_auto_total' });
+    let trans_public_total = store.createRecord('census-tract', { variable: 'trans_public_total' });
+
+    // Cannot get the Class constant import to work, so testing with magic strings
```
Comment (allthesignals) [inline frontend/tests/unit/models/census-tract-test.js]: `import { VARIABLE_MODE_LOOKUP } from 'labs-ceqr/app/models/census-tract';` should be 

`import { VARIABLE_MODE_LOOKUP } from 'labs-ceqr/models/census-tract';`

IDK why D:

File: frontend/app/templates/components/transportation/study-area-map/census-tract-popup/data.hbs [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<ThrottleProperty @property={{this.geoid}} as |geoid|>
+  <NeedsAsync @needs={{query "census-tract" (hash filter=(hash geoid=geoid))}} as |censusData|>
```
Comment (allthesignals) [inline frontend/app/templates/components/transportation/study-area-map/census-tract-popup/data.hbs]: cooooool, interested in your thoughts on this approach to using this addon. it is not something we have used before. I think it makes sense for these super ephemeral data fetches (that still go through the store).

**Comment by allthesignals** [comment]
Lemme know if I can help with the test situation:

> Ember Data Request GET /api/v1/census-tracts returned a 404

This seems weird to me...

File: frontend/tests/integration/components/transportation/study-area-map-test.js (hunk 6 above)
Comment (julialucyhogan) [reply frontend/tests/integration/components/transportation/study-area-map-test.js]: <3 googled for this for hrs lol

File: frontend/tests/unit/models/census-tract-test.js (hunk 8 above)
Comment (julialucyhogan) [reply frontend/tests/unit/models/census-tract-test.js]: i thought i tried that too... will give it another go

File: frontend/app/models/census-tract.js (hunk 4 above)
Comment (julialucyhogan) [reply frontend/app/models/census-tract.js]: Yes, but that other place is checking that there is a value for the population variable, not looking up  a human-readable translation of a variable name. Adding a crosswalk here does not get rid of the other 'Unknown' check.

**Comment by julialucyhogan** [comment]
Here are my thoughts RE `census-tract` as a name: it's wrong and bad. However, the data is not just modal split data either, because it includes other census information at the tract-level, specifically population. I think calling the data model in both front & backend `census-tract` is wrong, but I also think, given the way the data is structured calling it `modal-split` is wrong. What you actually are dealing with are single-geoid-census-variables but I cannot figure out what I want to call those.

**Comment by julialucyhogan** [comment]
> We should also use the CensusTract (or CensusTractRecord) Models in `models/transportation-analysis.jtwStudySelection` array.
> I.e. enforce objects in that array to be instances of the CensusTract(Record) Model.
> 
> Same for other analysis models (`public-schools`?) that use Census Tracts.

@godfreyyeung the reason we're not going to do this is b/c the census-tract models (to be renamed) are not a 1-1 relationship with the jtwStudySelection. Will be clearer when i've renamed it :)

**Comment by julialucyhogan** [comment]
closing this in favor of branch with correct issue in name

**Pull request closed without merging**